category_theory/natural_transformation.lean

```diff
@@ -0,0 +1,111 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Tim Baumann, Stephen Morgan, Scott Morrison
+
+import .functor
+
+open categories
+open categories.functor
+
+namespace categories.natural_transformation
+
+universes u₁ v₁ u₂ v₂ u₃ v₃
+
+section
+variable {C : Type u₁}
+variable [𝒞 : category.{u₁ v₁} C]
+variable {D : Type u₂}
+variable [𝒟 : category.{u₂ v₂} D]
+include 𝒞 𝒟
+
+structure NaturalTransformation (F G : C ↝ D) : Type (max u₁ v₂) :=
+  (components: Π X : C, (F +> X) ⟶ (G +> X))
+  (naturality: ∀ {X Y : C} (f : X ⟶ Y), (F &> f) ≫ (components Y) = (components X) ≫ (G &> f) . obviously)
+
+make_lemma NaturalTransformation.naturality
+attribute [ematch] NaturalTransformation.naturality_lemma
+
+infixr ` ⟹ `:50  := NaturalTransformation             -- type as \==> or ⟹
+
+definition IdentityNaturalTransformation (F : C ↝ D) : F ⟹ F := 
+{ components := λ X, 𝟙 (F +> X),
+  naturality := begin
+                  -- `obviously'` says:
+                  intros,
+                  simp
+                end }
+
+@[simp] lemma IdentityNaturalTransformation.components (F : C ↝ D) (X : C) : (IdentityNaturalTransformation F).components X = 𝟙 (F +> X) := by refl
+
+variables {F G H : C ↝ D}
+
+definition vertical_composition_of_NaturalTransformations (α : F ⟹ G) (β : G ⟹ H) : F ⟹ H := 
+{ components := λ X, (α.components X) ≫ (β.components X),
+  naturality := begin
+                  -- `obviously'` says:
+                  intros,
+                  simp,
+                  erw [←category.associativity_lemma, NaturalTransformation.naturality_lemma, category.associativity_lemma, ←NaturalTransformation.naturality_lemma]
+                end }
+
+notation α `⊟` β:80 := vertical_composition_of_NaturalTransformations α β    
+
+@[simp,ematch] lemma vertical_composition_of_NaturalTransformations.components (α : F ⟹ G) (β : G ⟹ H) (X : C) : (α ⊟ β).components X = (α.components X) ≫ (β.components X) := by refl
+
+-- We'll want to be able to prove that two natural transformations are equal if they are componentwise equal.
```

Comment:
It seems like this is an intermezzo between vertical and horizontal composition of natural transformations. Is that on purpose?

Reply:
No, it wasn't on purpose; I've rearranged now.